read-cache.c

```diff
@@ -1337,6 +1334,13 @@ static int add_index_entry_with_check(struct index_state *istate, struct cache_e
 	else
 		pos = index_name_stage_pos(istate, ce->name, ce_namelen(ce), ce_stage(ce));
 
+	/*
+	 * Cache tree path should be invalidated only after index_name_stage_pos,
+	 * in case it expands a sparse index.
+	 */
+	if (!(option & ADD_CACHE_KEEP_CACHE_TREE))
+		cache_tree_invalidate_path(istate, ce->name);
```

Comment:
I know we covered this yesterday, but would you mind reminding me why we invalidate certain cache tree paths?

Reply:
I _think_ this is what's happening, but I could be mistaken:

The cache tree represents the hierarchy of objects in the index as a series of trees and entries to those trees. Each tree "node" includes  subtree (i.e., subdirectory) information (pointers, subtree count), a _total_ number of elements attached to that tree (subdirectories and files), a name, and an object ID. The object ID here is a hash uniquely representing the full contents of a tree node/directory. 

When a file is modified in the index, the whole hierarchy of cache tree nodes for directories containing it (for a file at `folder1/folderA/myfile`, both `folderA` and `folder1`) has to change their object IDs, so we invalidate those entries. After all of the cache manipulation is done, calling `write_locked_index` internally calls `update_cache_tree`, reconstructing the contents of the tree with the new hashes.

Reply:
Thank you both for the excellent explanations!